simplify the usage of WebSocketUpgradeFilter in jetty 10

lachlan-roberts opened the pull request:
~~Now any `WebSocketUpgradeFilter` will use the default `WebSocketMapping` attribute by default. And we will not add a `WebSocketUpgradeFilter` automatically in the `JettyWebSocketServletContainerInitializer` if we detect there is already one added.~~

- Change `WebSocketUpgradeFilter` mapping init param name and default value.
- If the WebSocketMapping Attribute init parameter is not set on the `WebSocketUpgradeFilter` it will throw in its `init()` method.

lachlan-roberts committed:
simplify the usage of WebSocketUpgradeFilter

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented on jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java:
```diff
@@ -78,34 +77,40 @@
+    /**
+     * The init parameter name used to define {@link ServletContext} attribute used to share the {@link WebSocketMapping}.
+     */
+    public static final String MAPPING_ATTRIBUTE_INIT_PARAM = "org.eclipse.jetty.websocket.util.server.internal.WebSocketMapping.key";
```
This is the name of the init parameter used to define which `ServletContext` attribute to use to share the `WebSocketMapping`.

I am not really sure what use anyone would have for this as if they set it to anything else they would need to manually add their mappings to the `WebSocketMapping` which is an internal class.

I also don't really like the name `org.eclipse.jetty.websocket.util.server.internal.WebSocketMapping.key`, I think this could simplify this to just `WebSocketMapping`.

@gregw what do you think?

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-WebSocketUpgradeFilter

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/FilterHolder.java:
```diff
@@ -221,6 +222,39 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         }
     }
 
+    /**
+     * Work out the class of the held {@link Filter} even before the {@link FilterHolder} has been started.
+     * @return the class of the held {@link Filter}, or null if not available.
+     */
+    @SuppressWarnings("unchecked")
+    public Class<? extends Filter> getFilterClass()
+    {
+        if (_filter != null)
+            return _filter.getClass();
+
+        Filter filter = getInstance();
+        if (filter != null)
+            return filter.getClass();
+
+        Class<? extends Filter> heldClass = getHeldClass();
+        if (heldClass != null)
+            return heldClass;
+
+        String className = getClassName();
+        if (className != null)
+        {
+            try
+            {
+                return Loader.loadClass(className);
+            }
+            catch (ClassNotFoundException e)
+            {
+                LOG.warn("Could not load filter class", e);
+            }
+        }
+        return null;
+    }
+
```
I'm very cautious about this method!

firstly, this should probably be a mechanism on `BaseHolder` even if only used for filters.

Then if an instance has been set, then the heldClass will be set in the baseHolder, so you really can just look at `getHeldClass()` and only do something special if it is null. 

But my biggest concern is that the conversion of className to class is entirely dependent on the thread context classloader when called. This is well defined for when start is called, but not no define at all for when this method is called.... 

Need to think more about this... but please tidy up the above comments while I do!

gregw replied:
The other alternative is to use a very specific filter name.  Perhaps the fully qualified class name as the filter name?

gregw reviewed (changes requested):
I'm not really convinced this is workable as I'm not sure you can always convert a className into a class at the point you need to. 

They might extend the jetty servlet in their application, so the classname may be for something in the webapp, which might not be loadable until started.

Looking at the init param feels safer to me.

Let's discuss.

lachlan-roberts committed:
revert to using init parameter to identify the WebSocketUpgradeFilter

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Don't enforce the default WebSocketMapping.

If a user wants to use the WebSocketUpgradeFilter with a different WebSocketMapping,
they can add the ServerContainer manually and put their own WebSocketMapping in.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw reviewed (approved):
So this is really now just  a clean up and rename, rather than a substantive change.

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-WebSocketUpgradeFilter

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@gregw yes but also a change in behaviour, the `WebSocketUpgradeFilter` will now throw in its `init()` method if the mapping init parameter has not been set.

lachlan-roberts committed:
fix failures from incorrectly configured WebSocketUpgradeFilters

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>
